### DISCUSSION TITLE
Throttle next.js when building the page

### DISCUSSION THREAD

**Discussion by JustalK** [vercel/next.js#40131] [Help]
Hello Everyone,

We are currently using Next.js to build our website. Our website is connected to GraphCms that we use to fill up our static blog pages and guide pages. However, GraphCms has a rate limiter of 50 rps. When Next.js is building the page, the number of query sent by our website is way above this limit.

I was wondering, if there is a way to slow down the number of page being built, something like a rate limiter but for the builder of Next in order to not query GraphCms too suddenly ?

Thanks in advance.
Any help is appreciated :)

**Comment by icyJoseph** [comment]
Perhaps investigate some of these experimental options for `next.config.js` https://github.com/vercel/next.js/blob/481950c34b9c7b2eeb051eb4dc966d0ed3bb9605/packages/next/server/config-shared.ts#L91:

```ts
  cpus?: number
  sharedPool?: boolean
  profiling?: boolean
  isrFlushToDisk?: boolean
  workerThreads?: boolean
```

They are experimental though, but it could help to reduce the number of workers/cpus that build the pages. For example, if we model our understanding of the `cpus : 1`, option as that there'll be only one CPU building pages one after the other, then we could think that  the amount of requests per seconds has to dramatically fall. The trade-off would be that the build time would be larger.

**Reply by JustalK** [reply]
It tested and it does not slow down the building enough. I still reach the limit of GraphCms.

I was thinking, if I was using a different mode for the page such as Incremental static generation or server side rendering, would it work ? And in this case, how does the SEO work ?

**Reply by icyJoseph** [reply]
The SEO would be just fine. You could generate only the most important pages, to make sure that users see them directly. 

SEO, the way most people care about it, is only affected when the meta tags and head elements are not calculated on the server, at build or runtime. CSR pages struggle with this because initially there's only an app HTML shell that needs to be built with a runtime.

In some cases SEO crawlers don't run JS, so you'd also want the body elements to be calculated on the server. Both, pages without data fetching functions or with them, will generate some kind of HTML that's served on first request.

In your case, for the rest of least important pages, you can use `getStaticPaths` + `getStaticProps` with fallback true, or `getServerSideProps`, and SEO would be just fine.

Perhaps the one consideration is that `getServerSideProps` would run on every request, and might need an extra layer to cache for all users, and at the very least, have cache headers set per request, to throttle how much you reach out to your DB.

Using ISR with `getStaticPaths` and `getStaticProps`, might be the way to go, but still eventually there will be a number of pages you have to regenerate or revalidate all at once that'll trigger your limits, though this is less likely.

Alternatively an in-memory DB such as redis, could cache the CMS data briefly and that way you could totally throttle the amount of requests coming to it, independently of what the Next.js server is doing. This of course implies more costs and setup.
